### ISSUE TITLE
Exact does not work on goals with named metavariables

### ISSUE THREAD

**Issue by gebner** [leanprover/lean4#492]
### Prerequisites

* [x] Put an X between the brackets on this line if you have done all of the following:
    * Checked that your issue isn't already [filed](https://github.com/leanprover/lean4/issues).
    * Reduced the issue to a self-contained, reproducible test case.

### Description

This is about two issues regarding named metavariables that are somewhat interconnected:
 - Named metavariables cannot be assigned by unification after they have been created.
 - Exact does not work if the goal contains named metavariables.  (As well as other tactics that use the `ensureHasNoMVars` function.)

### Steps to Reproduce

It is often useful to fill in complicated witnesses using unification, however this does not work when using named metavariables:
```lean
example : ∃ n : Nat, n = n := by
  refine ⟨?n, ?h⟩
  case h => exact Eq.refl 3 -- type mismatch
```
I would expect that that the `exact` succeeds here and assigns the `?n` metavariable.

The `exact` tactic also fails for a different reason if `?n` does not need to be assigned:
```lean
example : ∃ n : Nat, n = n := by
  refine ⟨?n, ?h⟩
  case h => exact rfl -- don't know how to synthesize implicit argument
```
I would expect that `exact` succeeds here since it does not create any new metavariables.  After the exact I would expact a single goal that can be filled in with `case n => exact 3`.

The same issue also appears when using nested tactics:
```lean
example : ∃ n : Nat, n = n := by
  refine ⟨?n, by rfl⟩ -- don't know how to synthesize implicit argument
  case n => exact 3
```
I would expect the `rfl` to succeed in the nested tactic.

A related unexpected issue also occurs when using `refine` instead of `exact`:
```lean
example : ∃ n : Nat, n = n := by
  refine ⟨?n, ?h⟩
  case h =>
    refine rfl -- moves ?n goal into ?h
    case n => exact 3
```
I would expect that `refine rfl` closes the subgoal, and `case n => exact 3` succeeds afterwards (with the same indentation).

To summarize, I would expect that:
 - `exact` (and `rfl`, etc.) succeeds if it produces no *new* metavariables.
 - Named metavariables can be assigned by unification in subsequent tactics.
 - `refine` leaves existing goals intact if they are not assigned.

### Versions

Current nightly.

**Comment by leodemoura** [comment]
> - exact (and rfl, etc.) succeeds if it produces no new metavariables.
> - refine leaves existing goals intact if they are not assigned.

We will solve these two issues by adding a mechanism for detecting new metavariables created after a given point.

> Named metavariables can be assigned by unification in subsequent tactics.

I added this restriction to avoid a (fragile) hack that I added to Lean 3.
The goal is to make sure typing constraints resolved during elaboration should never "fill" holes that are supposed to be filled using tactics. 
That being said, I can see this is inconvenient in your example. So, I am considering the following modification.
The `exact t` tactic would elaborate `t` as it is done right now. So, during elaboration, we would still not fill named holes when solving typing constraints or TC resolution. Then, we ignore this restriction when we try to unify the type of `t` with the goal target type. The is defensible behavior since this last unification step is not really part of the term elaboration.

**Comment by gebner** [comment]
> The exact t tactic would elaborate t as it is done right now. So, during elaboration, we would still not fill named holes when solving typing constraints or TC resolution. Then, we ignore this restriction when we try to unify the type of t with the goal target type. The is defensible behavior since this last unification step is not really part of the term elaboration.

This should work for all examples that I can think of.
